streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredInternalProcessorContext.java

```diff
@@ -0,0 +1,312 @@
+        wrapped.setRecordContext(recordContext);
+    }
+
+    @Override
+    public void setCurrentNode(final ProcessorNode<?, ?, ?, ?> currentNode) {
+        wrapped.setCurrentNode(currentNode);
+    }
+
+    @Override
+    public ProcessorNode<?, ?, ?, ?> currentNode() {
+        return wrapped.currentNode();
+    }
+
+    @Override
+    public ThreadCache cache() {
+        return wrapped.cache();
+    }
+
+    @Override
+    public void initialize() {
+        wrapped.initialize();
+    }
+
+    @Override
+    public void uninitialize() {
+        wrapped.uninitialize();
+    }
+
+    @Override
+    public Task.TaskType taskType() {
+        return wrapped.taskType();
+    }
+
+    @Override
+    public void transitionToActive(final StreamTask streamTask, final RecordCollector recordCollector, final ThreadCache newCache) {
+        wrapped.transitionToActive(streamTask, recordCollector, newCache);
+    }
+
+    @Override
+    public void transitionToStandby(final ThreadCache newCache) {
+        wrapped.transitionToStandby(newCache);
+    }
+
+    @Override
+    public void registerCacheFlushListener(final String namespace, final ThreadCache.DirtyEntryFlushListener listener) {
+        wrapped.registerCacheFlushListener(namespace, listener);
+    }
+
+    @Override
+    public <T extends StateStore> T getStateStore(final StoreBuilder<T> builder) {
+        return wrapped.getStateStore(builder);
+    }
+
+    @Override
+    public void logChange(final String storeName, final Bytes key, final byte[] value, final long timestamp, final Position position) {
+        wrapped.logChange(storeName, key, value, timestamp, position);
+    }
+
+    @Override
+    public String changelogFor(final String storeName) {
+        return wrapped.changelogFor(storeName);
+    }
+
+    @Override
+    public void addProcessorMetadataKeyValue(final String key, final long value) {
+        wrapped.addProcessorMetadataKeyValue(key, value);
+    }
+
+    @Override
+    public Long processorMetadataForKey(final String key) {
+        return wrapped.processorMetadataForKey(key);
+    }
+
+    @Override
+    public void setProcessorMetadata(final ProcessorMetadata metadata) {
+        wrapped.setProcessorMetadata(metadata);
+    }
+
+    @Override
+    public ProcessorMetadata processorMetadata() {
+        return wrapped.processorMetadata();
+    }
+
+    public InternalProcessorContext<KOut, VOut> wrapped() {
+        return wrapped;
+    }
+
+    private class MeteredStateRestoreCallback implements StateRestoreCallback {
+
+        private final StateRestoreCallback wrapped;
+
+        MeteredStateRestoreCallback(final StateRestoreCallback wrapped) {
+            this.wrapped = wrapped;
+        }
+
+        @Override
+        public void restore(final byte[] key, final byte[] value) {
+            StreamsMetricsImpl.maybeMeasureLatency(() -> wrapped.restore(key, value), time, restoreSensor);
```

Comment:
The call to `restore()` does restore one "batch" of data, ie, whatever a single `poll()` call returned. So this is not the latency from "begin restore" to "complete restore".

We could maybe use `StateRestoreListener` callback, but also not sure if that's the best way?

I don't think we could use `maybeMeasureLatency()` helper method at all, but would need to get a ts when restore stars, and also when restore completes, and call `sensor.record(...)` directly?

This would go into StoreChangelogReader? Or we use existing `Task#recordRestoration()` method (or add similar overloads). We need to break it out per-store 🤔